community/values/src/main/java/org/neo4j/values/storable/UTF8StringValue.java

```diff
@@ -0,0 +1,445 @@
+            {
+                i++;
+            }
+
+            //The number of high bits tells us how many bytes we use to store the value
+            //e.g. 110xxxx -> need two bytes, 1110xxxx -> need three bytes, 11110xxx -> needs
+            //four bytes
+            while ( b < 0 )
+            {
+                i++;
+                b = (byte) (b << 1);
+            }
+            count++;
+        }
+        return count;
+    }
+
+    @Override
+    public int computeHash()
+    {
+        byte[] values = bytes;
+
+        if ( values.length == 0 || byteLength == 0 )
+        {
+            return 0;
+        }
+
+        int hash = 1, i = offset, len = offset + byteLength;
+        while ( i < len )
+        {
+            byte b = values[i];
+            //If high bit is zero (equivalent to the byte being positive in two's complement)
+            //we are dealing with an ascii value and use a single byte for storing the value.
+            if ( b >= 0 )
+            {
+                hash = 31 * hash + b;
+                i++;
+                continue;
+            }
+
+            //We can now have one of three situations.
+            //Byte1    Byte2    Byte3    Byte4
+            //110xxxxx 10xxxxxx
+            //1110xxxx 10xxxxxx 10xxxxxx
+            //11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+            //Figure out how many bytes we need by reading the number of leading bytes
+            int bytesNeeded = 0;
+            while ( b < 0 )
+            {
+                bytesNeeded++;
+                b = (byte) (b << 1);
+            }
+            int codePoint = codePoint( b, i, bytesNeeded );
+            i += bytesNeeded;
+
+            hash = 31 * hash + codePoint;
+        }
+
+        return hash;
+    }
+
+    @Override
+    public TextValue substring( int start, int length )
+    {
+        if ( start < 0 || length < 0)
+        {
+            throw new IndexOutOfBoundsException( "Cannot handle negative start index nor negative length" );
+        }
+
+        int end = start + length;
+        byte[] values = bytes;
+        int count = 0, byteStart = -1, byteEnd = -1, i = offset, len = offset + byteLength;
+        while ( i < len )
+        {
+            if ( count == start )
+            {
+                byteStart = i;
+            }
+            if ( count == end )
+            {
+                byteEnd = i;
+                break;
+            }
+            byte b = values[i];
+            //If high bit is zero (equivalent to the byte being positive in two's complement)
+            //we are dealing with an ascii value and use a single byte for storing the value.
+            if ( b >= 0 )
+            {
+                i++;
+            }
+
+            while ( b < 0 )
+            {
+                i++;
+                b = (byte) (b << 1);
+            }
+            count++;
+        }
+        if ( byteEnd < 0 )
+        {
+            byteEnd = len;
+        }
+        if (byteStart < 0 )
```

Comment:
The style god demands that you sacrifice a whitespace and add it after the opening paren on this line.